crates/rooch-executor/src/actor/reader_executor.rs

```diff
@@ -0,0 +1,626 @@
+    ListStatesMessage,
+};
+use anyhow::Result;
+use async_trait::async_trait;
+use coerce::actor::{context::ActorContext, message::Handler, Actor};
+use itertools::Itertools;
+use move_binary_format::errors::{Location, PartialVMError, VMResult};
+use move_core_types::account_address::AccountAddress;
+use move_core_types::identifier::{IdentStr, Identifier};
+use move_core_types::language_storage::ModuleId;
+use move_core_types::resolver::ModuleResolver;
+use move_core_types::value::MoveValue;
+use move_core_types::vm_status::{StatusCode, VMStatus};
+use move_resource_viewer::MoveValueAnnotator;
+use moveos::gas::table::{initial_cost_schedule, MoveOSGasMeter};
+use moveos::moveos::{GasPaymentAccount, MoveOS};
+use moveos::vm::vm_status_explainer::explain_vm_status;
+use moveos_store::transaction_store::TransactionStore;
+use moveos_store::MoveOSStore;
+use moveos_types::function_return_value::AnnotatedFunctionResult;
+use moveos_types::function_return_value::AnnotatedFunctionReturnValue;
+use moveos_types::module_binding::MoveFunctionCaller;
+use moveos_types::move_types::FunctionId;
+use moveos_types::moveos_std::event::EventHandle;
+use moveos_types::moveos_std::event::{AnnotatedEvent, Event};
+use moveos_types::moveos_std::tx_context::TxContext;
+use moveos_types::state::{AnnotatedState, State};
+use moveos_types::state_resolver::{AnnotatedStateReader, StateReader};
+use moveos_types::transaction::VerifiedMoveAction;
+use moveos_types::transaction::VerifiedMoveOSTransaction;
+use moveos_types::transaction::{FunctionCall, MoveAction};
+use moveos_types::transaction::{MoveOSTransaction, TransactionExecutionInfo};
+use moveos_verifier::metadata::load_module_metadata;
+use rooch_genesis::RoochGenesis;
+use rooch_store::RoochStore;
+use rooch_types::address::MultiChainAddress;
+use rooch_types::framework::address_mapping::AddressMapping;
+use rooch_types::framework::auth_validator::AuthValidatorCaller;
+use rooch_types::framework::auth_validator::TxValidateResult;
+use rooch_types::framework::transaction_validator::TransactionValidator;
+use rooch_types::framework::{system_post_execute_functions, system_pre_execute_functions};
+use rooch_types::transaction::AbstractTransaction;
+use rooch_types::transaction::AuthenticatorInfo;
+
+pub struct ReaderExecutorActor {
+    moveos: MoveOS,
+    rooch_store: RoochStore,
+}
+
+type ValidateAuthenticatorResult = Result<
+    (
+        TxValidateResult,
+        Option<MultiChainAddress>,
+        Vec<FunctionCall>,
+        Vec<FunctionCall>,
+    ),
+    VMStatus,
+>;
+
+impl ReaderExecutorActor {
+    pub fn new(
+        genesis: RoochGenesis,
+        moveos_store: MoveOSStore,
+        rooch_store: RoochStore,
+    ) -> Result<Self> {
+        let moveos = MoveOS::new(
+            moveos_store,
+            genesis.all_natives(),
+            genesis.config.clone(),
+            system_pre_execute_functions(),
+            system_post_execute_functions(),
+        )?;
```

Comment:
The writer executor needs to send a message to the reader executor after the transaction is executed and let the read executor update the state root.

Reply:
We can try and refactor it later.